sycl/include/syclcompat/device.hpp

```diff
@@ -508,31 +511,39 @@ Use 64 bits as memory_bus_width default value."
 #warning "get_device_info: querying the maximum number of \
     work groups is not supported."
 #endif
-    int max_nd_range_size[] = {0x7FFFFFFF, 0x7FFFFFFF, 0x7FFFFFFF};
-    prop.set_max_nd_range_size(max_nd_range_size);
+      int max_nd_range_size[] = {0x7FFFFFFF, 0x7FFFFFFF, 0x7FFFFFFF};
+      prop.set_max_nd_range_size(max_nd_range_size);
 #endif
 
-    // Estimates max register size per work group, feel free to update the value
-    // according to device properties.
-    prop.set_max_register_size_per_work_group(65536);
-
-    prop.set_global_mem_cache_size(
-        get_info<sycl::info::device::global_mem_cache_size>());
-
-    prop.set_image1d_max(get_info<sycl::info::device::image_max_buffer_size>());
-    prop.set_image1d_max(get_info<sycl::info::device::image_max_buffer_size>());
-    prop.set_image2d_max(get_info<sycl::info::device::image2d_max_width>(),
-                         get_info<sycl::info::device::image2d_max_height>());
-    prop.set_image3d_max(get_info<sycl::info::device::image3d_max_width>(),
-                         get_info<sycl::info::device::image3d_max_height>(),
-                         get_info<sycl::info::device::image3d_max_height>());
-    out = prop;
+      // Estimates max register size per work group, feel free to update the
+      // value according to device properties.
+      prop.set_max_register_size_per_work_group(65536);
+
+      prop.set_global_mem_cache_size(
+          get_info<sycl::info::device::global_mem_cache_size>());
+
+      prop.set_image1d_max(
+          get_info<sycl::info::device::image_max_buffer_size>());
+      prop.set_image1d_max(
+          get_info<sycl::info::device::image_max_buffer_size>());
+      prop.set_image2d_max(get_info<sycl::info::device::image2d_max_width>(),
+                           get_info<sycl::info::device::image2d_max_height>());
+      prop.set_image3d_max(get_info<sycl::info::device::image3d_max_width>(),
+                           get_info<sycl::info::device::image3d_max_height>(),
+                           get_info<sycl::info::device::image3d_max_height>());
+
+      _dev_info = prop;
+      out = prop;
+    } else {
+      out = *_dev_info;
+    }
   }
 
-  device_info get_device_info() const {
-    device_info prop;
-    get_device_info(prop);
-    return prop;
+  const device_info &get_device_info() const {
```

Comment:
Adding the `const` qualifier may be problematic. There could be code that uses our headers where:

```
prop = device.get_device_info();
prop.set_...();
```

It seems counterintuitive to modify a device property, but there may be some cases where this happens.
@joeatodd do you remember if this is the case?

Reply:
Hmm but in that case we'll need to drop the `const` qualifier of the `get_device_info()` as well? Or return by value which is not ideal. What do you think @Alcpz ?

Reply:
You mean the `const` at the end of the signature? Yes, you are right but it's a different issue. You are modifying the value of `_dev_info` in one of the paths.

Reply:
Hmm noted. I've switched back to the return-by-value to keep it safe for now.